control/velocity_controller/src/LQR_setup.cpp

```diff
@@ -0,0 +1,356 @@
+                                                   bool windup,
+                                                   double limit) {
+    if (abs(value) > limit) {
+        windup = true;
+        value = limit * (value / abs(value));
+    } else {
+        windup = false;
+    }
+    return {windup, value};
+}
+
+double LQRController::anti_windup(double ki,
+                                  double error,
+                                  double integral_sum,
+                                  bool windup) {
+    if (!windup) {
+        integral_sum += error * ki;
+    }
+    return integral_sum;
+}
+
+Eigen::Matrix3d LQRController::calculate_coriolis_matrix(double pitchrate,
+                                                         double yaw_rate,
+                                                         double sway_vel,
+                                                         double heave_vel) {
+    // Inertia matrix values??
+    Eigen::Matrix3d result;
+    result << 0.2, -30 * sway_vel * 0.01, -30 * heave_vel * 0.01,
+        30 * sway_vel * 0.01, 0, 1.629 * pitchrate, 30 * heave_vel * 0.01,
+        1.769 * yaw_rate, 0;
+    return result;
+}
+
+void LQRController::set_params(LQRparameters params) {
+    // set LQR parameters
+    integral_error_surge = 0.0;
+    integral_error_pitch = 0.0;
+    integral_error_yaw = 0.0;
+    surge_windup = false;
+    pitch_windup = false;
+    yaw_windup = false;
+    q_surge = params.q_surge;
+    q_pitch = params.q_pitch;
+    q_yaw = params.q_yaw;
+    r_surge = params.r_surge;
+    r_pitch = params.r_pitch;
+    r_yaw = params.r_yaw;
+    i_surge = params.i_surge;
+    i_pitch = params.i_pitch;
+    i_yaw = params.i_yaw;
+    i_weight = params.i_weight;
+    max_force = params.max_force;
+    return;
+}
+void LQRController::set_matrices(Eigen::Matrix3d inertia_matrix) {
+    inertia_matrix_inv = inertia_matrix.inverse();
+    state_weight_matrix.diagonal() << q_surge, q_pitch, q_yaw, i_weight,
+        i_weight, i_weight;
+    input_weight_matrix.diagonal() << r_surge, r_pitch, r_yaw;
+    return;
+}
+
+void LQRController::update_augmented_matrices(Eigen::Matrix3d coriolis_matrix) {
+    Eigen::Matrix3d system_matrix = inertia_matrix_inv;  //*coriolis_matrix;
+    // input_matrix = inertia_matrix_inv;
+    augmented_system_matrix << system_matrix(0, 0), system_matrix(0, 1),
+        system_matrix(0, 2), 0, 0, 0, system_matrix(1, 0), system_matrix(1, 1),
+        system_matrix(1, 2), 0, 0, 0, system_matrix(2, 0), system_matrix(2, 1),
+        system_matrix(2, 2), 0, 0, 0, -1, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0,
+        0, -1, 0, 0, 0;
+    augmented_input_matrix << inertia_matrix_inv(0, 0),
+        inertia_matrix_inv(0, 1), inertia_matrix_inv(0, 2), 0, 0, 0,
+        inertia_matrix_inv(1, 0), inertia_matrix_inv(1, 1),
+        inertia_matrix_inv(1, 2), 0, 0, 0, inertia_matrix_inv(2, 0),
+        inertia_matrix_inv(2, 1), inertia_matrix_inv(2, 2), 0, 0, 0;
```

Comment:
You should be able to initialize this with `::Zero()` then use `.topLeftCorner<>`, `.bottomLeftCorner<>` and `.leftCols<>` block operations to set the non-zero blocks. Arguably easier to read that way